woocommerce/woocommerce-admin
# Add currency extension

Fixes https://github.com/woocommerce/woocommerce-admin/issues/2345

Now that the [dataStore refactor](https://github.com/woocommerce/woocommerce-admin/pull/2961) introduced a way to add or modify SQL queries, some complex operation are possible for extensions. This PR aims to illustrate the addition of a UI element to introduce a query argument used to eventually modify SQL queries.

A currency viewer extension is introduced to Reports allowing the user to filter items by currency. Since currency data is part of post meta, an SQL join statement is required.

### Changes to wc-admin

* There were discrepancies between filter names using plural names, ie category v. categories, which were fixed.
* `persistedQueries` allow users to keep certain params in the url when navigating between reports. A filter was added.
* `getQueryFromConfig` was amended to use filter values from added configs.

### Screenshots

![Screen Shot 2019-12-02 at 3 09 33 PM](https://user-images.githubusercontent.com/1922453/69925582-d6c5b880-1515-11ea-8b2c-e7c60abd6527.png)

![Screen Shot 2019-12-03 at 2 55 38 PM](https://user-images.githubusercontent.com/1922453/70013578-0d1a3b00-15dd-11ea-95bc-0a908e07d5cf.png)

### Detailed test instructions:

1. Create two or more orders for today.
1. On the last order, change the postmeta value for key `_order_currency` to `NZD` or `ZAR`. If you are using vvv, you can go to the last page of the `wp_postmeta` table and edit the key value pair directly in phpmyadmin.
1. You may need to clear your cache `wp transient delete woocommerce_reports-transient-version`.
1. `npm run example -- --ext=sql-modification`
2. Activate WooCommerce Admin SQL modification Example
3. Go to any report and change the time to "today"
4. Manipulate the currency dropdown to see the results update accordingly.
5. Check all the reports.

### Follow up steps

* Clean up filter names with respect to pluralization. Also think about the catch-all hook names like `wc_admin_clauses_select` and see if one can be made for each report.
* The extension makes use of `wcSettings` by directly manipulating the global. It should be using `AssetDataRegistry`instead.
* When the currency dropdown is manipulated, currency symbols used in the chart and table should reflect that. I have a few ideas on this, but best to create a solution in the next PR.
